pkg/metrics/metrics_test.go

```diff
@@ -353,3 +353,141 @@ func TestWriteMetric(t *testing.T) {
 		})
 	}
 }
+
+func TestWriteMetricAsync(t *testing.T) {
+	t.Parallel()
+
+	// Record calls made to test server. Separate per test using a per-test
+	// unique id in URL.
+	var reqMap sync.Map
+
+	// Request body is intentionally leaked to allow for inspection in test cases.
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```

Comment:
I'm probably missing the point, but how about creating a separate test server for each subtest (inside the `for` loop) rather than sharing a server among all tests? That trivially guarantees isolation between the subtests.

Reply:
+1, was gonna suggest the same thing

Reply:
Yeah, that could be done. I was trying to reduce the overhead of starting test servers, but its probably somewhat trivial.

Reply:
Changed